[WIP] Add facility to get file details from patch server

john-tho opened the pull request:
Not ready for merge, only open for discussion

Currently only prints file details

Intention is that PyPoE could query and store Data details, and update files if need be: Query details, then when checksum fails, following update, only download the files that need updating.

Only tested on Linux

Documentation not tested and surely formatting errors

Wondering the best way to store file details?
After a quick look, it seems the [poe.file.ggpk records](https://github.com/OmegaK2/PyPoE/blob/3d608ddd825dc59e32c44c0ed9f86de2de250729/PyPoE/poe/file/ggpk.py#L855) are built around reading from an existing ggpk, and there is not currently the capability of creating the records from the information they will contain?

Could extend or adjust the poe.file.ggpk records to create virtual records, from the record data? Examples:
```
VirtualDirectoryRecord(name, hash)
VirtualFileRecord(name, size, hash)
```
Otherwise, create another set of classes?

john-tho committed:
Add patchserver PatchFileList and helpers to get file details from patch server

Add patchserver PatchFileList tests

Also adds patchserver utility classes to extend other PyPoE classes
BaseRecordData
VirtualDirectoryRecord
VirtualFileRecord
DirectoryNodeExtended

DirectoryNodeExtended methods
gen_walk(): generator
get_dict(): return dict of file|folder details
load_dict()

Example usage:

import PyPoE.poe.patchserver

patch = PyPoE.poe.patchserver.Patch()
patch_file_list = PyPoE.poe.patchserver.PatchFileList(patch)

patch_file_list.update_filelist(['Data'])

for node, depth in patch_file_list.directory.gen_walk():
  try:
    name = node.record.name
  except:
    name = 'ROOT'
  print('{blank:>{width}}{name}'.format(name=name, width=depth, blank=''))

from json import dump
dump_dict = patch_file_list.directory.get_dict()
dump_dict['version'] = patch_file_list.patch.version
file_handle = open('/tmp/poe.json', 'w', encoding='utf-8')
dump(dump_dict, file_handle)
file_handle.close()

import PyPoE.poe.patchserver
import json
from collections import OrderedDict

file_handle = open('/tmp/poe.json', 'rb')
file_dict = json.load(file_handle, object_pairs_hook=OrderedDict)
file_handle.close()
directory = PyPoE.poe.patchserver.DirectoryNodeExtended(None, None, None)
directory.load_dict(file_dict)

Chuanhsing commented:
python noob question, how to run PatchFileList

john-tho commented:
> how to run

Still very much work-in-progress
I have only tested on linux
After 6256792425b1d1cc32436e8681b9e769f8ccf032
open a python console (`ipython3` here) and:

```python
import PyPoE.poe.patchserver
patch = PyPoE.poe.patchserver.Patch()
patch_file_list = PyPoE.poe.patchserver.PatchFileList(patch)
patch_file_list.directory['Art']

def print_children(node, depth):
  if depth == 0:
    name = ''
  else:
    name = node.record.name
  print('{blank:>{width}}{name}'.format(name=name, width=depth, blank=''))

patch_file_list.directory.walk(print_children)

patch_file_list.update_filelist(['Art'])

patch_file_list.directory.walk(print_children)
```

Chuanhsing commented:
using other language, I got different response from "03 00 00", how to fix it?

your data
```
00000000:  04 00 00 00 00 00 17 00  00 0c 62 00 69 00 6e 00   |..........b.i.n.|
00000010:  6b 00 32 00 77 00 36 00  34 00 2e 00 64 00 6c 00   |k.2.w.6.4...d.l.|
00000020:  6c 00 00 05 aa 00 60 84  46 1a cd ff a3 05 01 bd   |l.....`.F.......|
00000030:  87 8b 01 65 c3 a9 5c 0e  2e ce cc 3f 16 5e 75 8f   |...e..\....?.^u.|
```

my data
```
00000000  04 00 00 04 a8 00 00 06  41 00 00 04 a8 50 00 00  |........A....P..|
00000010  00 00 17 05 00 f3 53 00  aa 05 00 0c 00 00 00 62  |......S........b|
00000020  00 69 00 6e 00 6b 00 32  00 77 00 36 00 34 00 2e  |.i.n.k.2.w.6.4..|
00000030  00 64 00 6c 00 6c 00 60  84 46 1a cd ff a3 05 01  |.d.l.l.`.F......|
```

john-tho commented:
@Chuanhsing 
> different response
> how to fix it?

Work out what the data means.
Or ignore it (skip over the parts you don't understand).

Is this on pathofexile.com, or another login server?

Different Endianness? Little?
`17` should be how many items this query returned
`05 00 f3 53` don't know, show me more data, or the patch server address?
One of these values show tell if the item is file or folder, or this might come later.
`00aa0500` is int size of the file (`371200=0x5aa00`)
`0c` is the length of the filename. This might be Little-Endian int and therefore  = `0000000c`
32 byte sha256sum is immediately after filename

Keep looking and work it out, then let me know.
Or let me know more details.

Chuanhsing commented:
using game launcher and wireshark get same result

attachment is port 12995.pcapng

[12995.zip](https://github.com/OmegaK2/PyPoE/files/1684106/12995.zip)

```
Attempting to connect to 'us.login.pathofexile.com 173.192.25.23' on port '12995'...OK.
00000000  01 05                                             |..|
00000002

00000000  02 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000010  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000020  00 00 28 68 00 74 00 74  00 70 00 3a 00 2f 00 2f  |..(h.t.t.p.:././|
00000030  00 70 00 61 00 74 00 63  00 68 00 63 00 64 00 6e  |.p.a.t.c.h.c.d.n|
00000040  00 2e 00 70 00 61 00 74  00 68 00 6f 00 66 00 65  |...p.a.t.h.o.f.e|
00000050  00 78 00 69 00 6c 00 65  00 2e 00 63 00 6f 00 6d  |.x.i.l.e...c.o.m|
00000060  00 2f 00 33 00 2e 00 31  00 2e 00 33 00 2e 00 33  |./.3...1...3...3|
00000070  00 2f 00 00 33 68 00 74  00 74 00 70 00 3a 00 2f  |./..3h.t.t.p.:./|
00000080  00 2f 00 70 00 61 00 74  00 63 00 68 00 35 00 30  |./.p.a.t.c.h.5.0|
00000090  00 32 00 2e 00 70 00 61  00 74 00 68 00 6f 00 66  |.2...p.a.t.h.o.f|
000000A0  00 65 00 78 00 69 00 6c  00 65 00 2e 00 63 00 6f  |.e.x.i.l.e...c.o|
000000B0  00 6d 00 3a 00 38 00 30  00 39 00 35 00 2f 00 70  |.m.:.8.0.9.5./.p|
000000C0  00 61 00 74 00 63 00 68  00 2f 00 33 00 2e 00 31  |.a.t.c.h./.3...1|
000000D0  00 2e 00 33 00 2e 00 33  00 2f 00                 |...3...3./.|
000000DB

00000000  03 00 00                                          |...|
00000003

00000000  04 00 00 04 a8 00 00 06  41 00 00 04 a8 50 00 00  |........A....P..|
00000010  00 00 17 05 00 f3 53 00  aa 05 00 0c 00 00 00 62  |......S........b|
00000020  00 69 00 6e 00 6b 00 32  00 77 00 36 00 34 00 2e  |.i.n.k.2.w.6.4..|
00000030  00 64 00 6c 00 6c 00 60  84 46 1a cd ff a3 05 01  |.d.l.l.`.F......|
00000040  bd 87 8b 01 65 c3 a9 5c  0e 2e ce cc 3f 16 5e 75  |....e..\....?.^u|
00000050  8f 5f 64 41 c3 ce c3 00  20 4f 8b 01 13 00 00 00  |._dA.... O......|
00000060  50 00 61 00 74 00 68 00  4f 00 66 00 45 00 78 00  |P.a.t.h.O.f.E.x.|
00000070  69 00 6c 00 65 00 5f 00  78 4f 00 f1 33 65 00 78  |i.l.e._.xO..3e.x|
00000080  00 65 00 2b 31 09 9e f8  0a e5 24 bf 70 8b eb b8  |.e.+1.....$.p...|
00000090  ac ff 7e f2 f8 6b 56 a0  de 4c dd 13 06 27 52 cd  |..~..kV..L...'R.|
000000A0  4d 3f 92 00 68 e6 34 00  12 00 00 00 64 00 33 00  |M?..h.4.....d.3.|
000000B0  64 00 63 00 6f 00 6d 00  70 00 5f 00 34 00 37 4d  |d.c.o.m.p._.4.7M|
000000C0  00 35 38 00 36 9c 00 f0  12 61 dd 8e 1f 6f 0c c0  |.58.6....a...o..|
000000D0  c1 4e 57 37 18 43 c0 86  67 4b 3b 7c a5 db 05 38  |.NW7.C..gK;|...8|
000000E0  1b b4 2e 02 6c 3f dd 86  86 01 e0 00 f5 37 03 00  |....l?.......7..|
000000F0  00 00 41 00 72 00 74 00  0a b8 3f ad c8 6b 3c b0  |..A.r.t...?..k<.|
00000100  ab 17 fa 8a 47 f9 bc c3  77 9d 09 66 f2 60 c7 71  |....G...w..f.`.q|
00000110  18 52 70 12 41 2e df da  00 00 8a 13 00 0e 00 00  |.Rp.A...........|
00000120  00 66 00 6d 00 6f 00 64  00 73 00 74 00 75 00 64  |.f.m.o.d.s.t.u.d|
00000130  00 69 00 6f 74 00 f4 12  bd 42 9a b6 82 e1 a2 f1  |.i.ot....B......|
00000140  57 22 a9 f6 c5 df b1 7f  99 6b 44 fb 48 cb 35 d3  |W".......kD.H.5.|
00000150  1d 19 c6 a8 b7 2a b0 8e  01 5a 01 f1 14 53 00 68  |.....*...Z...S.h|
00000160  00 61 00 64 00 65 00 72  00 43 00 61 00 63 00 68  |.a.d.e.r.C.a.c.h|
00000170  00 65 00 44 00 33 00 44  00 31 00 31 00 2e 00 70  |.e.D.3.D.1.1...p|
00000180  16 00 f1 17 6b 00 65 00  64 00 3e f2 ac ca cb d5  |....k.e.d.>.....|
00000190  3a 07 0e cd a8 41 98 0f  61 76 c1 08 ee d3 3c b5  |:....A..av....<.|
000001A0  9f 68 5c 33 44 06 4f a5  b4 27 57 00 1f 12 57 00  |.h\3D.O..'W...W.|
000001B0  06 0a 4d 00 f1 11 5c 00  4b 99 bb 02 06 56 e5 8f  |..M...\.K....V..|
000001C0  d0 2f aa dd b8 c0 e3 9c  71 8a 26 1c 2e 0f da b3  |./......q.&.....|
000001D0  ba 84 c8 03 15 5d 4d 00  51 07 00 00 00 6d f5 01  |.....]M.Q....m..|
000001E0  ff 1d 69 00 6d 00 61 00  70 00 1c b4 95 dc 0e 13  |..i.m.a.p.......|
000001F0  5f 7c ce a2 69 60 58 c5  34 cd 52 c5 10 3f 0b b4  |_|..i`X.4.R..?..|
00000200  0b cf b9 57 27 a4 a3 71  48 ba 00 68 b0 3f 9c 01  |...W'..qH..h.?..|
00000210  0a 08 38 02 f1 11 e9 94  84 7e 01 a6 f1 e4 cb dc  |..8......~......|
00000220  5a 86 46 16 ac 26 2f 67  ee 4f 14 db 19 49 84 66  |Z.F..&/g.O...I.f|
00000230  1a 8d 92 7a b7 f4 84 00  51 04 00 00 00 44 38 02  |...z....Q....D8.|
00000240  f3 19 61 00 29 4f 3e 55  40 e8 5b 38 c7 a4 0a 84  |..a.)O>U@.[8....|
00000250  fd e8 03 ad da 48 45 76  e4 17 0f 50 5b 02 30 2c  |.....HEv...P[.0,|
00000260  e9 8c 1e 34 00 20 49 0a  00 0d 69 02 00 ea 00 11  |...4. I...i.....|
00000270  43 fa 00 00 0a 00 04 5d  02 ff 17 8b 7f 77 e1 b5  |C......].....w..|
00000280  34 37 08 25 df f4 1d 07  81 26 b6 3c bf c3 8d e3  |47.%.....&.<....|
00000290  c7 52 bc 52 f0 9b 5a 9c  99 ad 51 00 20 c7 34 01  |.R.R..Z...Q. .4.|
000002A0  0f ac 02 06 04 47 00 fd  15 c5 00 a8 2c 2b 3f 54  |.....G......,+?T|
000002B0  dd ec b5 50 da 05 69 4d  a9 9f cb da 7d 8d af 9a  |...P..iM....}...|
000002C0  dc 1a 44 44 d7 24 7f f9  92 00 00 7a 04 34 03 35  |..DD.$.....z.4.5|
000002D0  33 00 32 fc 00 f7 17 ef  8a 08 95 04 d7 4a 30 14  |3.2..........J0.|
000002E0  a8 61 21 b2 83 18 59 2a  12 8f 07 45 20 24 2e 4a  |.a!...Y*...E $.J|
000002F0  52 d0 20 04 d6 7a a6 00  00 0c 1b 00 0a 69 02 08  |R. ..z.......i..|
00000300  39 01 f5 11 54 fb 3c ec  e0 10 c0 a3 09 f3 7a ea  |9...T.<.......z.|
00000310  36 58 1a d8 d1 8e fe 8a  65 1f 3d cc 19 7d c1 3c  |6X......e.=..}.<|
00000320  85 18 fa 8a bd 01 08 0a  02 f1 17 73 00 b1 47 0f  |...........s..G.|
00000330  68 0f 4d 3a b7 9a 8a 27  17 13 44 25 38 05 16 a4  |h.M:...'..D%8...|
00000340  e3 92 56 dc ef 01 26 f0  81 e5 93 09 11 00 20 8f  |..V...&....... .|
00000350  07 74 00 b5 43 00 6c 00  69 00 65 00 6e 00 74 f2  |.t..C.l.i.e.n.t.|
00000360  00 f1 11 1e 92 d3 35 61  f6 03 a1 4d 90 a7 83 7a  |......5a...M...z|
00000370  35 15 ad 40 d0 b7 10 60  ba 73 35 90 13 80 70 74  |5..@...`.s5...pt|
00000380  47 27 a1 74 00 11 05 49  03 04 10 03 f7 17 c4 f3  |G'.t...I........|
00000390  7b 95 7b 43 fe 16 2e 94  e7 5e 3c 22 84 8c 9c f1  |{.{C.....^<"....|
000003A0  18 eb 76 b6 bf ea af 8b  d0 c9 70 15 2a e6 00 00  |..v.......p.*...|
000003B0  a8 16 00 08 e4 00 04 e0  00 f1 11 59 40 16 4b 3a  |...........Y@.K:|
000003C0  f7 a4 57 d6 ca 7b d0 df  91 90 ce 77 a9 b4 cc e5  |..W..{.....w....|
000003D0  fe 85 35 84 d2 4f 0b 35  72 d1 bb 6c 00 00 39 00  |..5..O.5r..l..9.|
000003E0  31 4d 00 65 19 02 13 64  21 02 f5 11 27 7c e1 71  |1M.e...d!...'|.q|
000003F0  0d 0a 4f 6b ab 31 19 fa  20 3f 3b 87 b7 30 48 89  |..Ok.1.. ?;..0H.|
00000400  f5 00 d5 f2 4f db d0 d0  a2 c3 eb 26 52 02 f1 19  |....O......&R...|
00000410  46 00 4d 00 4f 00 44 00  8d 94 01 97 16 7f 2b df  |F.M.O.D.......+.|
00000420  50 a4 52 2c a2 9c 14 0d  77 0c 69 7f 92 3c 19 85  |P.R,....w.i..<..|
00000430  c9 33 0f 4f c6 b8 ff 2c  31 00 37 11 00 00 48 03  |.3.O...,1.7...H.|
00000440  9b 64 00 48 00 4c 00 53  00 4c 5e 01 ff 11 c6 39  |.d.H.L.S.L^....9|
00000450  d1 ef 51 0d 9f 02 0d fd  7f 13 df 8d 88 94 58 0e  |..Q...........X.|
00000460  9d 09 cf 9f 6b fa 1e bb  c1 00 31 03 82 25 1b 03  |....k.....1..%..|
00000470  04 02 58 02 13 72 23 03  0f 1b 03 15 6f 00 00 52  |..X..r#.....o..R|
00000480  1a 00 10 88 04 04 08 2b  02 ff 12 13 2d 5b 53 9e  |.......+....-[S.|
00000490  44 76 84 cc 10 58 53 e1  c1 8c 87 a9 72 60 94 7f  |Dv...XS.....r`..|
000004A0  6d 9c 5d 12 49 67 c4 8d  0a e9 76 00 01 00 43 50  |m.].Ig....v...CP|
000004B0  00 00 00 00 00                                    |.....|
000004B5
Closing socket...OK.
```

john-tho commented:
Patch server responds differently if it has state for the connection.
Client does get webroot `0105` (Client log says `Send patching protocol version 5`), get root `030000` on the one connection.

Starting the connection with patching protocol version 4 `0104`, gets the easier interpreted data, but there is slightly more data.

This python code closes connection between the `0105` and `030000`. Without proto 5 announced for the connection, patch server responds in proto 4.

Time to look into proto 5 I guess.

Chuanhsing commented:
I got the problem, check logs/Client.txt
[Info Client 4276] Send patching protocol version 5

_PROTO = b'\x01\x05' means
01: get patch update_patch_urls
05: version 5


when I use \x01\x04, the result will same as your program.

OmegaK2 commented:
This looks pretty cool. Are you going to make further changes? I'd wait with the pull in that case

john-tho commented:
@OmegaK2 
> further changes

Yes, further changes planned:

-  Check a passed file against what it should be: size, sha256sum.
-  Download a requested file if it does not compare:
   -  Probably in files / folders to start with, but then I lose directory hash
   -  Possibly export patch server file list information?
   -  Could nearly build a ggpk with only requested files? Do we know the murmur2 seed?

Also, how to you generate your sphinx docs, so I can fix my notes?

Any other tips or changes?

I would like to use patch proto 5 (what the launcher currently uses), but haven't figured it all out yet.
It seems to use for each item:
(Little-Endian int file size
LE int name length
name)  but for most entries these values are truncated or compressed?
sha256sum for all entries

OmegaK2 commented:
1. Go to docs subfolder 
2. Run python generate_rst_templates.py
3. Run make html

I'm not 100% sure the docs are fine at the moment, I don't build them very often and don't have tests for those setup yet.

john-tho committed:
Add patchserver use single proto 4 connection to server

Add patchserver patch socket sharing tests

Process:
open socket
tell server proto 4
detach from socket, set instance.sock_fd

When Patch instance deleted:
shutdown (FIN) patchserver connection
close patchserver socket

john-tho committed:
Add patchserver node_check_hash function

Add patchserver node_check_hash tests

Example GGPKFile:

from PyPoE.poe.file.ggpk import GGPKFile
ggpk_file = '/mnt/poe/Path_of_Exile/Content.ggpk'
ggpk = GGPKFile()
ggpk.read(ggpk_file)
ggpk.directory_build()

from PyPoE.poe.patchserver import node_check_hash
node_check_hash(ggpk['Data/Maps.dat'], ggpk=ggpk)

test_node = 'Art/2DArt/Atlas'

node = ggpk[test_node]

node_hashes = node_check_hash(node, ggpk=ggpk)

if node_hashes[-1][2] is True:
  print('hashes match for {}'.format(node.get_path()))
else:
  for child_node, child_hash, child_bool in node_hashes:
    if bool is False:
      print(child_node.record.name)
      print('{} file hash'.format(child_hash.hexdigest()))
      print('{:064x} expected hash'.format(child_node.record.hash))

import PyPoE.poe.patchserver
patch = PyPoE.poe.patchserver.Patch()
patch_file_list = PyPoE.poe.patchserver.PatchFileList(patch)
cdir = ''
for dir in test_node.split('/'):
  if cdir:
      cdir += '/'
  cdir += dir
  patch_file_list.update_filelist([cdir])

from hmac import compare_digest
node_patchserver_results = []
for index, child in enumerate(node_hashes):
    node_path = child[0].get_path()
    node_patchserver = patch_file_list.directory[node_path]
    node_file_hash = child[1]
    hash_test = compare_digest(
        node_file_hash.hexdigest(),
        format(node_patchserver.record.hash, '064x'))

    node_patchserver_results.append(hash_test)

if list(zip(node_hashes, node_patchserver_results))[-1][1] is True:
  print('hashes match patchserver for {}'.format(test_node))

Example files:

import os
import PyPoE.poe.patchserver
from PyPoE.poe.patchserver import VirtualFileRecord

patch = PyPoE.poe.patchserver.Patch()
patch_file_list = PyPoE.poe.patchserver.PatchFileList(patch)
poe_dir = '/mnt/poe/Path_of_Exile'
for node, depth in patch_file_list.directory.gen_walk(max_depth=1):
    if (node.record is not None and
        isinstance(node.record, VirtualFileRecord)):
            node_hash = node_check_hash(node, folder_path=poe_dir)
            print('{} hashed okay?: {}'.format(
                node.record.name,
                node_hash[0][2]))

A tale of finding PyPoE DirectoryNode children folder hash permutation:

from PyPoE.poe.file.ggpk import GGPKFile
ggpk_file = '/mnt/poe/Path_of_Exile/Content.ggpk'
ggpk = GGPKFile()
ggpk.read(ggpk_file)
ggpk.directory_build()

from PyPoE.poe.patchserver import node_check_hash
node = ggpk['Art/2DArt/Atlas']
node_hashes = node_check_hash(node, ggpk=ggpk)

import itertools
from hmac import compare_digest
from hashlib import sha256

found = False
perms = itertools.permutations(node_hashes[0:len(node_hashes)-1])
for index, perm in enumerate(perms):
  # known that ggpk hashes over network are lowercase hex bytes,
  # and that folder hashes are concat of item hashes
  folder_hash_concat = b''
  for child in list(perm):
    folder_hash_concat += child[1].digest()
  folder_hash = sha256(folder_hash_concat)
  if compare_digest(folder_hash.hexdigest(), format(node.record.hash, '064x')):
    found = True
    perms_match_index = index
    break

if found == True:
    print('match found: perms[{}]'.format(perms_match_index))
    print('### hash match order ###:')
    for child in list(perm):
      print(child[0].record.name)

print('### PyPoE ggpk directory children order ###')
for child in node.children:
  print(child[0].record.name)

john-tho commented:
I cleaned this up a bit and had a go at fixing my documentation.

Any changes you would like to see?

I have not tested it on Windows. I think this should be done before merge.

Do you think I should add a function to update selected node files in an on-disk directory?
Best to also add extra metadata to the folders, so that folder hash and file order for hashing are known?
Otherwise, build a GGPK file for selected nodes?

eps1lon commented:
@john-tho any instructions to what and how I should test this on Windows?

john-tho committed:
Add patchserver Patch.download_raw iterate through patch file hosts on connection refused

john-tho committed:
Fix patchserver Patch.download_raw url cut unneeded slash

Both patch_cdn_url and patch_url end with a trailing forward slash

john-tho committed:
Reformat patchserver test rename url to file

john-tho committed:
Fix patchserver test use pytest tmpdir

john-tho commented:
I added some pytests for my changes.

If you run `py.test tests/PyPoE/poe/test_patchserver.py` and it passes, this should (if I got my tests right) mean you can:

- connect to patchserver `patch = Patch()`
- download game file from CDN `patch.download(...)`
- download game file metadata from patchserver `PatchFileList(patch)`
- confirm game file matches expected patchserver hash `node_check_hash(...)`

For actual usage, there should be examples in the commit messages (9557583) and source  (for docs).

eps1lon commented:
Seems to work. Nice job 👍

Windows 10 (Version=1709, OS Build=16299.248)
```
$ py.test tests/PyPoE/poe/test_patchserver.py
============================= test session starts =============================
platform win32 -- Python 3.4.0, pytest-3.3.2, py-1.5.2, pluggy-0.6.0
rootdir: C:\Development\src\js\poe\poedb\pypoe, inifile:
collected 10 items

tests\PyPoE\poe\test_patchserver.py ..........                           [100%]
========================== 10 passed in 2.51 seconds ==========================
```

john-tho committed:
Add patchserver node_outdated_files

Example:

import PyPoE.poe.patchserver
patch = PyPoE.poe.patchserver.Patch()
patch_file_list = PyPoE.poe.patchserver.PatchFileList(patch)
poe_dir = '/tmp/pypoe/poe'
node = 'Data'
update_list = PyPoE.poe.patchserver.node_files_outdated(
    patch_file_list, node, poe_dir, recurse=False)

john-tho committed:
Add patchserver node_update_files

Example:

import PyPoE.poe.patchserver
patch = PyPoE.poe.patchserver.Patch()
patch_file_list = PyPoE.poe.patchserver.PatchFileList(patch)
poe_dir = '/tmp/pypoe/poe'
node = 'Metadata/Items/Amulets'
PyPoE.poe.patchserver.node_update_files(
    patch_file_list, node, poe_dir, recurse=False)
node = 'Art/2DArt/Atlas'
PyPoE.poe.patchserver.node_update_files(
    patch_file_list, node, poe_dir, recurse=False)
